src/protocol/thread/thread.ts

```diff
@@ -1047,28 +1093,51 @@ class _ThreadFront {
     return this._chooseSourceIdList(this.getSourceIdsForURL(url));
   }
 
-  // Given the sourceId of a preferred or alternate source and its URL,
-  // get all sourceIds of preferred/alternate sources with the same URL.
-  getCorrespondingSourceIds(sourceId: SourceId, url: string | undefined) {
-    if (!url) {
-      return [sourceId];
-    }
+  private groupSourceIds() {
+    for (const [sourceId, source] of this.sources.entries()) {
+      if (!source.url) {
+        this.correspondingSourceIds.set(sourceId, [sourceId]);
+        continue;
+      }
+      if (this.correspondingSourceIds.has(sourceId)) {
+        continue;
+      }
+
+      const groups = this.getChosenSourceIdsForUrl(source.url);
+      assert(groups.length > 0);
+      const sourceIds = groups.map(group => group.sourceId);
+      for (const sourceId of sourceIds) {
+        this.correspondingSourceIds.set(sourceId, sourceIds);
+      }
+      const alternateIds = groups
+        .map(group => group.alternateId)
+        .filter(alternateId => !!alternateId) as SourceId[];
```

Comment:
If you don't want to do ` as SourceId[];` you can also do

    .filter((alternateId): alternateId is SourceId => !!alternateId);

I believe

Reply:
Nice, thanks.